crates/provider/src/heart.rs

```diff
@@ -458,37 +488,74 @@ impl<S> Heartbeat<S> {
 }
 
 #[cfg(target_arch = "wasm32")]
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
+impl<S: Stream<Item = Block> + Unpin + 'static, N: Network> Heartbeat<S, N> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
 #[cfg(not(target_arch = "wasm32"))]
-impl<S: Stream<Item = Block> + Unpin + Send + 'static> Heartbeat<S> {
+impl<S: Stream<Item = Block> + Unpin + Send + 'static, N: Network> Heartbeat<S, N> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
-    async fn into_future(
+type TransportReceipt<R> = TransportResult<Option<R>>;
+
+/// Spawns a receipt fetching task. This task receives transaction hashes and fetches receipts for
+/// requested transactions, sending the results back.
+fn spawn_receipt_fetcher<T: Transport + Clone, N: Network>(
+    client: WeakClient<T>,
+) -> ReceiptFetcherHandle<N> {
+    let (receipt_tx, receipt_rx) = mpsc::channel(16);
+    let (hash_tx, mut hash_rx) = mpsc::channel(16);
```

Comment:
ah okay, I get how this works.

this spawns a dedicated fetcher task that receives the receipts spawns eth_getTransactionReceipt requests